cryopose/_validators.py

```diff
@@ -1,24 +1,53 @@
+from typing import Union, Sequence, Literal
+
+import numpy as np
+from numpy.typing import ArrayLike, NDArray
 import pandas as pd
+from pandas.api.types import is_numeric_dtype, is_string_dtype
+from scipy.spatial.transform import Rotation
 
 from ._data_labels import CryoPoseDataLabels as CPDL
 
 
-def validate_3d_cryopose_df(df: pd.DataFrame) -> bool:
-    """Validate a cryopose dataframe for particle poses in 3D."""
-    required_columns_3d = [
-        CPDL.POSITION_X,
-        CPDL.POSITION_Y,
-        CPDL.POSITION_Z,
-        CPDL.ORIENTATION,
-    ]
-    return all(k in df for k in required_columns_3d)
-
-
-def validate_2d_cryopose_df(df: pd.DataFrame) -> bool:
-    """Validate a cryopose dataframe for particle poses in 3D."""
-    required_columns_2d = [
-        CPDL.POSITION_X,
-        CPDL.POSITION_Y,
-        CPDL.ORIENTATION,
-    ]
-    return all(k in df for k in required_columns_2d)
+def validate_positions(positions: ArrayLike, ndim: Literal[2, 3] = 3) -> NDArray:
+    positions = np.asarray(positions, dtype=float)
+    if positions.ndim != 2 or positions.shape[1] not in (2, 3):
+        raise ValueError(f'positions must be a (n, 2) or (n, 3) array, got {positions.shape}')
+    if ndim != positions.shape[1]:
+        raise ValueError(f'positions are {positions.shape[1]}D, but a {ndim}D cryopose was requested')
+    return positions
+
+
+def validate_orientations(orientations: Union[Rotation, Sequence[Rotation]], ndim: Literal[2, 3] = 3) -> Rotation:
+    orientations = Rotation.concatenate(orientations)
+    if ndim == 2:
+        # TODO: check that rotation is only inplane?
+        ...
```

Comment:
2D particles can still have 3D rotations :)

Need to decide on semantics for what a 3D orientation for a 2D particle means

Reply:
3D tomogram space doesn't really exist for single particle so the semantics can't be exactly the same :)